kibana-reports/server/routes/savedSearchReportHelper.ts

```diff
@@ -0,0 +1,180 @@
+      metaData.paternName = resIndexPattern.title;
+      (metaData.timeFieldName = resIndexPattern.timeFieldName),
+        (metaData.fields = resIndexPattern.fields); // Get all fields
+      // Getting fields of type Date
+      for (const item of JSON.parse(metaData.fields)) {
+        if (item.type === 'date') {
+          metaData.dateFields.push(item.name);
+        }
+      }
+    }
+  }
+}
+
+async function generateReport(client: IClusterClient | IScopedClusterClient) {
+  const report = { _source: metaData };
+
+  const dataset: any = [];
+  const arrayHits: any = [];
+  let esData: any = {};
+
+  // fetch ES query max size windows
+  const indexPattern: string = report._source.paternName;
+  const settings = await client.callAsInternalUser('indices.getSettings', {
+    index: indexPattern,
+    includeDefaults: true,
+  });
+  const defaultMaxSize: number =
+    settings[indexPattern].settings.index.max_result_window != null
+      ? settings[indexPattern].settings.index.max_result_window
+      : settings[indexPattern].defaults.index.max_result_window;
+
+  // build the ES Count query
+  const countReq = buildQuery(report, 1);
+  // Count the Data in ES
+  const esCount = await client.callAsInternalUser('count', {
+    index: indexPattern,
+    body: countReq.toJSON(),
+  });
+
+  // If No data in elasticsearch
+  const total = esCount.count;
+  if (total === 0) {
+    return {};
+  }
+
+  // build the ES query
+  const reqBody = buildQuery(report, 0);
+
+  if (total > defaultMaxSize) {
+    // fetch the data
+    esData = await fetchData(report, reqBody, defaultMaxSize);
+    arrayHits.push(esData.hits);
+
+    const nbScroll = Math.floor(total / defaultMaxSize);
+
+    for (let i = 0; i < nbScroll - 1; i++) {
+      const resScroll = await client.callAsInternalUser('scroll', {
+        scrollId: esData._scroll_id,
+        scroll: '1m',
+      });
+      if (Object.keys(resScroll.hits.hits).length > 0) {
+        arrayHits.push(resScroll.hits);
+      }
+    }
+    const extraFetch = total % defaultMaxSize;
+    if (extraFetch > 0) {
+      const extraEsData = await fetchData(report, reqBody, extraFetch);
+      arrayHits.push(extraEsData.hits);
+    }
+  } else {
+    esData = await fetchData(report, reqBody, total);
+    arrayHits.push(esData.hits);
+  }
+
+  // Get data
+  dataset.push(getEsData(arrayHits, report));
+  // Convert To csv
+  return await convertToCSV(dataset);
+
+  // Fetch the data from ES
+  async function fetchData(report, reqBody, fetchSize) {
+    const docvalues = [];
+    for (const dateType of report._source.dateFields) {
+      docvalues.push({
+        field: dateType,
+        format: 'date_hour_minute',
+      });
+    }
+
+    const newBody = {
+      query: reqBody.toJSON().query,
+      docvalue_fields: docvalues,
+    };
+
+    return await client.callAsInternalUser('search', {
+      index: report._source.paternName,
+      scroll: '1m',
```

Comment:
I am not sure what is 1m scroll means. is it 1 million rows? if so is this require to be that big?

Reply:
I just refactored it to a constant. This is actually the expire time for scroll context. Thanks.